### ISSUE TITLE
Freezing Model Layers Directly in HUB

### ISSUE THREAD

**Issue by AK-algos** [ultralytics/hub#939]
### Search before asking

- [X] I have searched the HUB [issues](https://github.com/ultralytics/hub/issues) and [discussions](https://github.com/ultralytics/hub/discussions) and found no similar questions.

### Question

Is there a way to freeze model layers during training directly in Ultralytics HUB?

### Additional

I have a model that has already been trained on 7000 images, I would like to now introduce 1000 net new images and pick up training where I left off with the old model (without regressing). Is there a way to freeze model layers directly in Ultralytics HUB when running the training? I see the option to run training on an existing model but no option for explicitly freezing.

**Comment by pderrenger** [comment]
Hello! Thank you for your thoughtful question.

At the moment, freezing specific model layers during training is not directly configurable within the Ultralytics HUB user interface. However, you can still achieve this by leveraging HUB's ability to export and fine-tune models locally, where you would have full control over the training configuration, including freezing layers.

Here’s how you can proceed:

1. **Export Your Model from HUB:**
   Download your existing trained model (with the 7000 images) from [Ultralytics HUB](https://hub.ultralytics.com). Ensure you select the model file (`.pt` format) that you wish to continue training on.

2. **Set Up Local Training with Frozen Layers:**
   Use the `--freeze` argument in the training script (`train.py`) to specify which layers to freeze. For example:
   ```
   python train.py --weights your_model.pt --data new_data.yaml --epochs 50 --freeze 10
   ```
   Here, `--freeze 10` freezes the first 10 layers (e.g., the backbone), as seen in the [YOLOv5 freezing guide](https://docs.ultralytics.com/yolov5/tutorials/transfer_learning_with_frozen_layers/). You can adjust the number based on the layers you wish to freeze.

3. **Reintroduce the Refined Model to HUB:**
   Once local training is complete, you can re-upload your updated model back into your HUB project for further deployment or evaluation.

### Additional Notes
Freezing layers is particularly useful for transfer learning tasks like yours, where you want to preserve learned knowledge from an earlier dataset while refining the model on new data. Keep in mind that freezing too many layers might limit the model's adaptability to new features in the additional images.

If you'd like this feature integrated directly into HUB, we encourage you to suggest it through the [Ultralytics HUB GitHub Issues](https://github.com/ultralytics/hub/issues). The community and team are always looking to enhance HUB's capabilities based on user feedback.

Let me know if you need any further clarification or assistance setting this up! 🚀

**Comment by AK-algos** [comment]
Thank you for the quick response, this is very helpful. Just wanted to confirm one additional element that I found in the Ultralytics HUB UI. Screenshot below, based on what you mentioned this feature directly would not work at this time, right?

<img width="823" alt="Screenshot 2024-12-01 at 8 43 44 AM" src="https://github.com/user-attachments/assets/746add45-a726-4020-af97-e4c912efed71">

**Comment by sergiuwaxmann** [comment]
@AK-algos You can use the Custom field present in the Advanced Model Configuration.
If you set `freeze` to `10` it will force `--freeze 10` - which is what you need.

**Comment by AK-algos** [comment]
Amazing thank you!

**Comment by sergiuwaxmann** [comment]
@AK-algos Glad we could help!

By the way, this section from our documentation relates to your question:

![SCR-20241201-pvp](https://github.com/user-attachments/assets/2bfb9f51-2e29-4bbb-8c38-ae2f08f69faa)
https://docs.ultralytics.com/hub/models#2-model

**Comment by AK-algos** [comment]
Sounds good! One related question - since my model is already trained on the original 7000 images, would it be advised to freeze the backbone and run additional training on only the incremental 1000 images or include all 8000? I've heard the former can sometimes lead to model degradation but wasn't sure if freezing the backbone would alleviate that issue.

**Comment by pderrenger** [comment]
That's a great question! The decision whether to include all 8000 images or only the incremental 1000 depends on your specific goals and the characteristics of the new data, but let me provide some insights:

### Option 1: Freeze the Backbone and Train on 1000 New Images
Freezing the backbone is a common strategy when incremental data is relatively small (as in your case) and similar in distribution to the original dataset. It prevents the model from "unlearning" features it already learned from the 7000 images and focuses the training effort on fine-tuning the detection head (which adjusts to new class distributions or scenarios in the new dataset).

However, the risk of degradation arises if the new data introduces significantly different features or classes that the backbone needs to learn to detect effectively. If the 1000 images include fundamentally different scenarios or conditions, freezing could limit the model's ability to adapt.

### Option 2: Train on All 8000 Images
Retraining on the combined 8000 images ensures the model can learn harmoniously from both the old and new datasets. While this approach requires more computational resources, it mitigates the risk of regression by allowing the backbone to adjust if necessary. You don't have to explicitly freeze or unfreeze layers here.

### Recommendation
Given your situation:
- If the new 1000 images are similar to the original 7000 and cover the same problem domain, freezing the backbone and training only on the incremental data should suffice.
- If there's a domain shift (e.g., new classes, diverse backgrounds), retraining on all 8000 images without freezing would likely yield better results.

### What to Watch Out For
1. **Overfitting on the Smaller Dataset**: If you train using only 1000 images, there’s an increased risk of overfitting. Regularization strategies like data augmentation or dropout will help alleviate this.
2. **Longer Training Time on 8000 Images**: If computation time is a concern, consider implementing stratified sampling to proportionally include a mix of original and incremental data for reduced training size.

### Practical Implementation in Ultralytics HUB
If you're using Ultralytics HUB:
- **To Train on 8000 Images**: Simply merge the datasets and re-upload to HUB. Start training with your existing model as the base weights.
- **To Train with Frozen Backbone on 1000 Images**: Use the "Custom" field under advanced options in the HUB UI, and specify `freeze=10` to freeze the backbone. Train with only the additional dataset.

A mixed approach might also work well, such as including a subset of 7000 images (reducing training size) with the new 1000 images and training with or without freezing.

If you want to experiment further, HUB makes it super convenient to run multiple training configurations and compare performance metrics to find the optimal strategy. Let me know if you'd like additional clarifications or tips! 🚀

**Comment by AK-algos** [comment]
Thank you for the comprehensive response, makes sense! Appreciate it

**Comment by pderrenger** [comment]
You're very welcome! 😊 I'm glad the explanation helped clarify things for you. If you have any additional questions as you proceed—whether they're about freezing layers, dataset configurations, or anything else related to training and deploying your YOLO models in Ultralytics HUB—feel free to reach out here.

Remember, experimenting with different setups (like training with or without the frozen backbone) and tracking performance metrics in HUB can help fine-tune your process. Wishing you the best results with your model! 🚀
